core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -270,27 +327,91 @@ class LogSegment(val log: FileMessageSet,
     catch {
       case e: IOException => throw kafkaStorageException("index", e)
     }
+    try timeIndex.renameTo(new File(CoreUtils.replaceSuffix(timeIndex.file.getPath, oldSuffix, newSuffix)))
+    catch {
+      case e: IOException => throw kafkaStorageException("timeindex", e)
+    }
+  }
+
+  /**
+   * Append the last time index entry to the time index when this log segment become inactive segment.
+   * This entry will be used to decide when to delete the segment.
+   */
+  def onBecomeInactiveSegment() {
+    if (maxTimestampSoFar >= 0)
+      timeIndex.maybeAppend(maxTimestampSoFar, offsetOfMaxTimestamp, skipFullCheck = true)
+    else
+      timeIndex.maybeAppend(lastModified, nextOffset() - 1, skipFullCheck = true)
+  }
+
+  /**
+   * The time this segment has waited to be rolled over. If the segment does not contain any message with a timestamp,
+   * the time is based on the create time of the segment. Otherwise the time is based on the max timestamp in this
+   * segment.
+   */
+  def timeWaitedForRoll(now: Long) = now - { if (maxTimestampSoFar >= 0) maxTimestampSoFar else created}
+
+  /**
+   * Search the message offset based on timestamp.
+   * This method returns a TimestampOffset. The offset is the offset of the first message whose timestamp is
+   * greater than or equals to the target timestamp.
+   *
+   * If all the message in the segment have smaller timestamps, the offset will be last offset + 1 and the
+   * timestamp will be max timestamp in the segment.
```

Comment:
Hmm, if none of the message has timestamp, should we include the offset of the first message in the segment?